importer/export.hs

```diff
@@ -0,0 +1,155 @@
+-- but WITHOUT ANY WARRANTY; without even the implied warranty of
+-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+-- Lesser General Public License for more details.
+--
+-- You should have received a copy of the GNU Lesser General Public
+-- License along with this library; if not, see <http://www.gnu.org/licenses/>.
+
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE RecordWildCards #-}
+
+import           Control.Monad(when)
+import           Control.Monad.Except(MonadError, runExceptT, throwError)
+import           Control.Monad.IO.Class(MonadIO, liftIO)
+import           Control.Monad.Trans.Resource(MonadBaseControl, MonadResource, runResourceT)
+import           Data.ByteString(ByteString)
+import           Data.Conduit((.|), Producer, runConduit, yield)
+import           Data.Conduit.Binary(sinkFile)
+import qualified Data.Conduit.List as CL
+import           Data.Maybe(fromMaybe, listToMaybe)
+import           Data.Text(Text, pack, unpack)
+import           Data.Time.Clock.POSIX(posixSecondsToUTCTime)
+import           Database.Esqueleto
+import           Database.Persist.Sqlite(runSqlite)
+import           System.Directory(createDirectoryIfMissing, createFileLink, setModificationTime)
+import           System.Environment(getArgs)
+import           System.Exit(exitFailure)
+import           System.FilePath((</>), dropDrive, takeDirectory)
+import           System.Posix.Files(setFileMode)
+import           System.Posix.Types(CMode(..))
+
+import GI.Gio    hiding(on)
+import GI.GLib   hiding(String, on)
+import GI.OSTree hiding(on)
+
+import qualified BDCS.CS as CS
+import           BDCS.DB
+import           BDCS.Files(groupIdToFiles)
+
+-- Convert a GInputStream to a conduit source
+sourceInputStream :: (MonadResource m, IsInputStream i) => i -> Producer m ByteString
+sourceInputStream input = do
+    let buf_size = 8096
+    bytes <- liftIO $ inputStreamReadBytes input buf_size noCancellable
+    bytesSize <- liftIO $ bytesGetSize bytes
+    if bytesSize == 0
+        then return ()
+        else do
+            bytesData <- liftIO $ bytesGetData bytes
+            yield $ fromMaybe "" bytesData
+            sourceInputStream input
+
+nameToGroupId :: MonadIO m => Text -> SqlPersistT m (Maybe (Key Groups))
+nameToGroupId name = do
+    ndx <- select $ distinct $ from $ \(keyval `InnerJoin` group_keyval `InnerJoin` groups) -> do
+           on     $ keyval ^. KeyValId ==. group_keyval ^. GroupKeyValuesKey_val_id &&.
+                    group_keyval ^. GroupKeyValuesGroup_id ==. groups ^. GroupsId
+           where_ $ keyval ^. KeyValKey_value ==. val "name" &&.
+                    keyval ^. KeyValVal_value ==. val name &&.
+                    groups ^. GroupsGroup_type ==. val "rpm"
+           limit 1
+           return $ groups ^. GroupsId
+    return $ listToMaybe (map unValue ndx)
+
```

Comment:
Sure, that'd make sense in BDCS.Groups. I guess what kept me from putting it there was avoiding questions like how else might we want to map a thing to a group id, or how to resolve multiple group id returned for a thing. Maybe I could turn that into something rpm-specific and return the one with highest EVR, or just require NEVRA as input or something.